### ISSUE TITLE
computation and data transfer could not be overlapping

### ISSUE THREAD

**Issue by jeng1220** [cupy/cupy#1938]
Hi,
I found that computation and data transfer could not be overlapping in CuPy.
All CUDA commands were serialized.
However, using CUDA C, the same behavior was overlapping.

* Conditions
CuPy Version : 5.1.0  
CUDA Build Version : 10000  
CUDA Driver Version : 10000  
CUDA Runtime Version : 10000  
GPU : GTX1080TI  
OS : Ubuntu 16.04

* Code to reproduce
CuPy version:
```python
import numpy as np
import cupy as cp
from optparse import OptionParser

pinned_memory_pool = cp.cuda.PinnedMemoryPool()
cp.cuda.set_pinned_memory_allocator(pinned_memory_pool.malloc)
def pin_memory(array):
    mem = cp.cuda.alloc_pinned_memory(array.nbytes)
    ret = np.frombuffer(mem, array.dtype, array.size).reshape(array.shape)
    ret[...] = array
    return ret

def multistream_test(nb_task, mat_size, nb_stream):
  x_pin = pin_memory(np.ones((mat_size, mat_size), dtype=np.float32))
  y_pin = pin_memory(np.ones((mat_size, mat_size), dtype=np.float32))
  z_pin = pin_memory(np.ones((nb_task, mat_size, mat_size), dtype=np.float32)) 
  streams = []
  xs_gpu = []
  ys_gpu = []
  zs_gpu = []

  for stream_id in range(nb_stream):
    streams.append(cp.cuda.stream.Stream())
    xs_gpu.append(cp.empty_like(x_pin))
    ys_gpu.append(cp.empty_like(y_pin))
    zs_gpu.append(cp.empty_like(z_pin[0]))

  for i in range(nb_task):
    stream_index = i % nb_stream
    stream = streams[stream_index]
    x_gpu = xs_gpu[stream_index]
    y_gpu = ys_gpu[stream_index]
    z_gpu = zs_gpu[stream_index]

    stream.use()
    x_gpu.set(x_pin)
    y_gpu.set(y_pin)
    cp.dot(x_gpu, y_gpu, z_gpu)
    z_pin[i] = z_gpu.get()

  for i in range(nb_stream):
    streams[i].synchronize()

def main():
  parser = OptionParser()
  parser.add_option('-s', '--size', dest='size',
                    default=1000,
                    help='matrix size')
  parser.add_option('-q', '--queue', dest='queue',
                    default=4,
                    help='number of stream')
  parser.add_option('-t', '--task', dest='task',
                    default=32,
                    help='number of task')
  (options, args) = parser.parse_args()

  nb_stream = int(options.queue)

  nb_task = int(options.task)
  remain = nb_task % nb_stream
  if remain != 0:
    padding = nb_stream - remain
    nb_task += padding
    print('padding number of task to be multiple of amount of stream', 
      'new number of task:', nb_task)

  mat_size = int(options.size)

  multistream_test(nb_task, mat_size, nb_stream)
  multistream_test(nb_task, mat_size, nb_stream)

if __name__ == '__main__':
  main()
```

CUDA C version:
```c
#include <cuda_runtime.h>
#include <cublas_v2.h>

void multistream_test()
{
  const int nb_stream = 4;

  cudaStream_t streams[nb_stream];
  float* x_pin;
  float* y_pin;
  float* z_pin;
  float* xs_gpu[nb_stream];
  float* ys_gpu[nb_stream];
  float* zs_gpu[nb_stream];

  const int mat_size = 1000;
  const int nb_task = 36;
  const size_t mat_size_inbyte = sizeof(float) * mat_size * mat_size;
  cudaHostAlloc((void**)&x_pin, mat_size_inbyte, cudaHostAllocDefault);
  cudaHostAlloc((void**)&y_pin, mat_size_inbyte, cudaHostAllocDefault);
  cudaHostAlloc((void**)&z_pin, mat_size_inbyte * nb_task, cudaHostAllocDefault);
  for (int i = 0; i < nb_stream; ++i) {
    cudaStreamCreate(&streams[i]);
    cudaMalloc((void**)&xs_gpu[i], mat_size_inbyte);
    cudaMalloc((void**)&ys_gpu[i], mat_size_inbyte);
    cudaMalloc((void**)&zs_gpu[i], mat_size_inbyte);
  }

  cublasHandle_t handle;
  cublasCreate(&handle);

  for (int i = 0; i < nb_task; ++i) {

    int stream_index = i % nb_stream;
    cudaStream_t stream = streams[stream_index];
    float* x_gpu = xs_gpu[stream_index];
    float* y_gpu = ys_gpu[stream_index];
    float* z_gpu = zs_gpu[stream_index];

    cudaMemcpyAsync(x_gpu, x_pin, mat_size_inbyte, cudaMemcpyHostToDevice, stream);
    cudaMemcpyAsync(y_gpu, y_pin, mat_size_inbyte, cudaMemcpyHostToDevice, stream);
    cublasSetStream(handle, streams[stream_index]);
    float alpha = 1.f;
    float beta  = 0.f;
    cublasSgemm(handle,
                CUBLAS_OP_N, CUBLAS_OP_N,
                mat_size, mat_size, mat_size,
                &alpha,
                x_gpu, mat_size,
                y_gpu, mat_size,
                &beta,
                z_gpu, mat_size);
    size_t offset = i * mat_size * mat_size;
    cudaMemcpyAsync(z_pin + offset, z_gpu, mat_size_inbyte, cudaMemcpyDeviceToHost, stream);
  }

  for (int i = 0; i < nb_stream; ++i) {
    cudaStreamSynchronize(streams[i]);
  }

  cudaFreeHost(x_pin);
  cudaFreeHost(y_pin);
  cudaFreeHost(z_pin);
  for (int i = 0; i < nb_stream; ++i) {
    cudaStreamDestroy(streams[i]);
    cudaFree(xs_gpu[i]);
    cudaFree(ys_gpu[i]);
    cudaFree(zs_gpu[i]);
  }

  cublasDestroy(handle);
}


int main(int argc, const char** argv)
{
  multistream_test();
  return 0;
}
```

build CUDA C sample:
```sh
$ nvcc multistream_test.c -lcublas -o multistream_test.out
```

* Error messages, stack traces, or logs
nvvp logs:
CuPy [link](https://drive.google.com/file/d/1MKNJiuBKeRFEtY44cbHP4XYIB2B3Neef/view?usp=sharing)
CUDA C [link](https://drive.google.com/file/d/1xslXopbOcMOA4TKl1_0R4Rsco2h1rYhH/view?usp=sharing)

**Comment by grafi-tt** [comment]
I think this issue can be closed as #1955 is merged.

**Comment by jeng1220** [comment]
Yes, using new `get(out)` API can solve this issue. Close issue.
